### DISCUSSION TITLE
Is lazy evaluation possible for a query key

### DISCUSSION THREAD

**Discussion by vmichailidisdev** [TanStack/query#6751] [Q&A] [answered]
Hello,
So, I want to fetch something but I only know the key that will be assigned to the cache after the fetch is finished.

The only way I can think is:
```javascript
fetch(url).then(data => queryClient.setQueryData(createKeyFrom(data), data))
```

Is there any cleaner way of doing this? (I don't mind having the same data under 2 query keys.)

**Comment by TkDodo** [comment]
> but I only know the key that will be assigned to the cache after the fetch is finished.

how is that possible? The key is the dependency to the fetch ...

**Reply by vmichailidisdev** [reply]
Quick response as always, thank you Dominik. I will try to be as clear as I can.

Let's say we have ui that we can search for users. Furthermore, we can click a button to send a message to a user. So, we can cache a query like ['conversations', userId] with the conversation data.

Then, we have another ui where we can see our latest messages. There we fetch conversation data according to the conversationId (['conversations', conversationId]). So, we cannot leverage the already fetched data that are under ['conversations', userId].

It will be perfect to cache the same data under 2 or more query keys. I am just trying to find a clean way of doing this.

**Reply by TkDodo** [reply]
> So, we cannot leverage the already fetched data

that's right, but I don't know how / why "changing" the queryKey would be a solution to this problem. I would maybe pre-fill the `['conversations', userId]` cache with data from the `['conversations', conversationId]` cache if it exists. For example, we can call `queryClient.getQueriesData(['conversations'])` to get all conversations and then loop through them to see if there is one that matches our `conversationId`. This can be one in `initialData`, and if we don't have any, the queryFn will just run as usual...

**Reply by vmichailidisdev** [reply]
Thank you very much for the suggestion. It will fit just fine.

**Reply by vmichailidisdev** [reply]
What about listening to cache events and when data with this specific query key are added, add the data under another query key?

**Reply by TkDodo** [reply, marked as answer]
sure, you can do `queryClient.getQueryCache().subscribe` and listen to all events that come in and do with those what you like. That's also what the devtools do :)
